trustfall_core/test_data/tests/parse_errors/variable_in_root_query.graphql.ron

```diff
@@ -0,0 +1,11 @@
+TestGraphQLQuery (
+    schema_name: "numbers",
+    query: r#"{
+  query Number($max: Int!) {
```

Comment:
@obi1kenobi I can't seem to figure out how to write a graphql query to test this 😅 
```
query HeroForEpisode($ep: Episode!) {
  hero(episode: $ep) {
    name
    ... on Droid {
      primaryFunction
    }
    ... on Human {
      height
    }
  }
}
```
Using the following example ^ This will throw a `MultipleOperationsInDocument` before it can throw a `VariableDefinitionInRootQuery` and I can't seem to find a way to use variables in graphql outside of having a top level query definition like so. I'm assuming trustfall uses graphql but as a different dialect where this is not allowed in the first place for queries.

Reply:
Huh. I'm confused too, I don't understand what the multiple definitions are. That sounds like a bug to me, and it's possibly in my understanding of some GraphQL core concept.

Could you `Debug` print the multiple definitions and put them here?

I would also expect that query to trigger the `VariableDefinitionInRootQuery` error case in the parser. Feel free to tweak the logic to make it so, unless the multiple definitions thing seems to indicate we'd be making a mistake doing that.

Reply:
```
{Name("MyQuery"): Positioned { pos: Pos(2:1), node: OperationDefinition { ty: Query, variable_definitions: [Positioned { pos: Pos(2:15), node: VariableDefinition { name: Positioned { pos: Pos(2:16), node: Name("max") }, var_type: Positioned { pos: Pos(2:21), node: Type { base: Named(Name("Int")), nullable: false } }, directives: [], default_value: None } }], directives: [], selection_set: Positioned { pos: Pos(2:27), node: SelectionSet { items: [Positioned { pos: Pos(3:5), node: Field(Positioned { pos: Pos(3:5), node: Field { alias: None, name: Positioned { pos: Pos(3:5), node: Name("Number") }, arguments: [(Positioned { pos: Pos(3:12), node: Name("max") }, Positioned { pos: Pos(3:17), node: Variable(Name("max")) })], directives: [], selection_set: Positioned { pos: Pos(3:23), node: SelectionSet { items: [Positioned { pos: Pos(4:9), node: Field(Positioned { pos: Pos(4:9), node: Field { alias: Some(Positioned { pos: Pos(4:9), node: Name("num") }), name: Positioned { pos: Pos(4:14), node: Name("value") }, arguments: [], directives: [Positioned { pos: Pos(4:20), node: Directive { name: Positioned { pos: Pos(4:21), node: Name("output") }, arguments: [] } }], selection_set: Positioned { pos: Pos(0:0), node: SelectionSet { items: [] } } } }) }] } } } }) }] } } } }}
```

This is whats debug printing for the match case where its throwing `MultipleOperationsInDocument` error.  From the graphql async parser: 

```
The operations of a GraphQL document.
There is either one anonymous operation or many named operations.
```

Is the expectation that trustfall should only support anonymous operations? Or that named operations should be parsed out? It's the parse library thats lexing this as multiple operations.

Reply:
Could you debug print `document` as a whole, and use pretty-printed output `{:#?}` to make it easier to read?

I can't tell exactly which part of the query this printed output is, and it's quite difficult to read too because it's all on one gigantic line.

I don't think we have a stance on named vs anonymous operations. I just don't see multiple operations in here at all, and I'm wondering whether that's a bug upstream or a bug in my understanding of the GraphQL spec.

Reply:
Sorry my bad 😆 - I'll go ahead and pretty print it. 
```
{
    Name(
        "MyQuery",
    ): Positioned {
        pos: Pos(2:1),
        node: OperationDefinition {
            ty: Query,
            variable_definitions: [
                Positioned {
                    pos: Pos(2:15),
                    node: VariableDefinition {
                        name: Positioned {
                            pos: Pos(2:16),
                            node: Name(
                                "max",
                            ),
                        },
                        var_type: Positioned {
                            pos: Pos(2:21),
                            node: Type {
                                base: Named(
                                    Name(
                                        "Int",
                                    ),
                                ),
                                nullable: false,
                            },
                        },
                        directives: [],
                        default_value: None,
                    },
                },
            ],
            directives: [],
            selection_set: Positioned {
                pos: Pos(2:27),
                node: SelectionSet {
                    items: [
                        Positioned {
                            pos: Pos(3:5),
                            node: Field(
                                Positioned {
                                    pos: Pos(3:5),
                                    node: Field {
                                        alias: None,
                                        name: Positioned {
                                            pos: Pos(3:5),
                                            node: Name(
                                                "Number",
                                            ),
                                        },
                                        arguments: [
                                            (
                                                Positioned {
                                                    pos: Pos(3:12),
                                                    node: Name(
                                                        "max",
                                                    ),
                                                },
                                                Positioned {
                                                    pos: Pos(3:17),
                                                    node: Variable(
                                                        Name(
                                                            "max",
                                                        ),
                                                    ),
                                                },
                                            ),
                                        ],
                                        directives: [],
                                        selection_set: Positioned {
                                            pos: Pos(3:23),
                                            node: SelectionSet {
                                                items: [
                                                    Positioned {
                                                        pos: Pos(4:9),
                                                        node: Field(
                                                            Positioned {
                                                                pos: Pos(4:9),
                                                                node: Field {
                                                                    alias: Some(
                                                                        Positioned {
                                                                            pos: Pos(4:9),
                                                                            node: Name(
                                                                                "num",
                                                                            ),
                                                                        },
                                                                    ),
                                                                    name: Positioned {
                                                                        pos: Pos(4:14),
                                                                        node: Name(
                                                                            "value",
                                                                        ),
                                                                    },
                                                                    arguments: [],
                                                                    directives: [
                                                                        Positioned {
                                                                            pos: Pos(4:20),
                                                                            node: Directive {
                                                                                name: Positioned {
                                                                                    pos: Pos(4:21),
                                                                                    node: Name(
                                                                                        "output",
                                                                                    ),
                                                                                },
                                                                                arguments: [],
                                                                            },
                                                                        },
                                                                    ],
                                                                    selection_set: Positioned {
                                                                        pos: Pos(0:0),
                                                                        node: SelectionSet {
                                                                            items: [],
                                                                        },
                                                                    },
                                                                },
                                                            },
                                                        ),
                                                    },
                                                ],
                                            },
                                        },
                                    },
                                },
                            ),
                        },
                    ],
                },
            },
        },
    },
}
```

Reply:
FYI this is the following document:
```
query MyQuery($max: Int!) {
    Number(max: $max) {
        num: value @output
    }
}
```

Normally when I use graphQL we don't rely on anonymous operations so more often then not its formatted like the above ^^ or like VVV
```
mutation Mutation($vars: Type!) {
	operation(foo: $vars) {
		returnVal
	}
}
```

Reply:
When printing out the `&document` value before this match statement https://github.com/obi1kenobi/trustfall/blob/f94cab5bd26e64ac095d2d4242edd10b9c9b2c6e/trustfall_core/src/graphql_query/query.rs#L126

I'm seeing:
```
ExecutableDocument {
    operations: Multiple(
        {
            Name(
                "MyQuery",
            ): Positioned {
                pos: Pos(2:1),
                node: OperationDefinition {
                    ty: Query,
                    variable_definitions: [
                        Positioned {
                            pos: Pos(2:15),
                            node: VariableDefinition {
                                name: Positioned {
                                    pos: Pos(2:16),
                                    node: Name(
                                        "max",
                                    ),
                                },
                                var_type: Positioned {
                                    pos: Pos(2:21),
                                    node: Type {
                                        base: Named(
                                            Name(
                                                "Int",
                                            ),
                                        ),
                                        nullable: false,
                                    },
                                },
                                directives: [],
                                default_value: None,
                            },
                        },
                    ],
                    directives: [],
                    selection_set: Positioned {
                        pos: Pos(2:27),
                        node: SelectionSet {
                            items: [
                                Positioned {
                                    pos: Pos(3:5),
                                    node: Field(
                                        Positioned {
                                            pos: Pos(3:5),
                                            node: Field {
                                                alias: None,
                                                name: Positioned {
                                                    pos: Pos(3:5),
                                                    node: Name(
                                                        "Number",
                                                    ),
                                                },
                                                arguments: [
                                                    (
                                                        Positioned {
                                                            pos: Pos(3:12),
                                                            node: Name(
                                                                "max",
                                                            ),
                                                        },
                                                        Positioned {
                                                            pos: Pos(3:17),
                                                            node: Variable(
                                                                Name(
                                                                    "max",
                                                                ),
                                                            ),
                                                        },
                                                    ),
                                                ],
                                                directives: [],
                                                selection_set: Positioned {
                                                    pos: Pos(3:23),
                                                    node: SelectionSet {
                                                        items: [
                                                            Positioned {
                                                                pos: Pos(4:9),
                                                                node: Field(
                                                                    Positioned {
                                                                        pos: Pos(4:9),
                                                                        node: Field {
                                                                            alias: Some(
                                                                                Positioned {
                                                                                    pos: Pos(4:9),
                                                                                    node: Name(
                                                                                        "num",
                                                                                    ),
                                                                                },
                                                                            ),
                                                                            name: Positioned {
                                                                                pos: Pos(4:14),
                                                                                node: Name(
                                                                                    "value",
                                                                                ),
                                                                            },
                                                                            arguments: [],
                                                                            directives: [
                                                                                Positioned {
                                                                                    pos: Pos(4:20),
                                                                                    node: Directive {
                                                                                        name: Positioned {
                                                                                            pos: Pos(4:21),
                                                                                            node: Name(
                                                                                                "output",
                                                                                            ),
                                                                                        },
                                                                                        arguments: [],
                                                                                    },
                                                                                },
                                                                            ],
                                                                            selection_set: Positioned {
                                                                                pos: Pos(0:0),
                                                                                node: SelectionSet {
                                                                                    items: [],
                                                                                },
                                                                            },
                                                                        },
                                                                    },
                                                                ),
                                                            },
                                                        ],
                                                    },
                                                },
                                            },
                                        },
                                    ),
                                },
                            ],
                        },
                    },
                },
            },
        },
    ),
    fragments: {},
}
```

Reply:
Taking a look at the source code for the graphQL parsing library this is where its defining the enum that gets returned:
https://docs.rs/async-graphql-parser/latest/src/async_graphql_parser/parse/executable.rs.html#30

Reply:
Based on reading the GraphQL spec, it seems like `Single` vs `Multiple` is a misnomer and better names might be `SingleAnonymous` vs `Named`.

With that in mind, I feel our best course of action would be to leave a comment describing this weirdness (for the sake of future readers of this code), and then write a bunch of test cases to pin down both our upstream library's behavior and our own.

We should accept single named operations and run them normally without error. We should not accept multiple named operations — that's where we want to return the "multiple operations" error, I think.

Reply:
Okay this sounds good. So something like this 
```
query Query() {
	Number(max: 3) {
		num: value @output
	}
}
```

should be fine but something like this 

```
query Query() {
	Number(max: 3) {
		num: value @output
	}
	Number(max: 8) {
		num: value @output
	}
}
```

should error?

Reply:
I think the latter should error for a different reason -- it has the same output name twice, but there's still only one named operation.

```graphql
query Query1() {
  Number(max: 3) {
    num: value @output
  }
}

query Query2() {
  Number(max: 3) {
    value @output
  }
}
```
This has two named operations, and should error with the "multiple operations" error.